### PR TITLE
fix: correctness bug sweep across cubic

### PR DESCRIPTION
**Pull request by alxndrkalinin**
## Summary

A review-driven sweep of correctness bugs across the codebase, each fix paired with a regression test. 10 atomic commits, grouped by concept.

Full suite passes on the **A40 GPU** (`uv run pytest`: 354 passed, 0 skipped — every GPU-gated test exercised), plus `ruff check`/`ruff format` and `mypy` clean (pre-commit hooks ran on each commit).

## Fixes

- **image_utils** — `crop_corner("br")` returned the whole image (reachable via `crop_br` in the FRC local-resolution sweep); `pad_image_to_shape` fell one short on odd size differences and tripped its assert (broke `richardson_lucy_xp` with mismatched PSFs). Now both correct and `crop_center`-consistent.
- **preprocessing** — `decon_*` 3D-only slicing (`, :, :`) raised on 2D; made rank-agnostic. `richardson_lucy_skimage` observer path re-clipped each iteration and fed the deconvolution back as the image — now recomputes per depth so snapshots are true *i*-iteration results and the final matches the no-observer call.
- **segmentation** — `remove_touching_objects` used an additive `border_value=100` offset that misclassified any label id > 100 as touching, silently deleting non-touching objects (Cellpose emits hundreds of labels). Now reads neighbours directly. `clear_border` raised a tuple (`TypeError`) instead of the intended `ValueError`.
- **cuda / proxies** — SciPy & scikit-image proxies detected device from only the first arg; a GPU array under any other keyword routed to CPU and crashed. Added `is_gpu_array`, scan all args, coerce on the CPU branch.
- **metrics/ms_ssim** — kernel width now tracks `kernel_size` (was desynced from the Gaussian radius, leaking `cval=0` for small kernels); `(N,H,W)` now reduces per image (was pooling across the batch). 2D unchanged.
- **metrics/average_precision** — count distinct labels instead of `.max()` so supplied non-sequential matches give correct FP/FN/AP.
- **metrics/feature** — `get_true_features` matched base feature names against expanded column names, raising spuriously for vector properties.
- **metrics/spectral/radial** — `use_max_nyquist` was ignored for index units, compressing the XY-sector DCR/FSC curve on anisotropic shapes (default `dcr_resolution` path).
- **feature/mesh** — `extract_features` included background label 0; now excluded, matching `voxel.regionprops`.
- **plot_utils** — `show_image_error` color limits were asymmetric; now centered on zero.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by Sourcery

Fix a range of correctness issues across image utilities, deconvolution wrappers, GPU-aware SciPy/skimage proxies, and evaluation metrics, adding targeted regression tests for each case.

Bug Fixes:
- Correct crop_corner corner selection so all four directional helpers return the intended image region and align padding with crop_center in pad_image_to_shape, including odd size differences.
- Make Richardson–Lucy deconvolution wrappers rank-agnostic for 2D/3D inputs and fix the skimage observer mode so per-iteration snapshots and the final result match the no-observer call.
- Fix remove_touching_objects to detect touching labels via direct neighborhood inspection instead of a border offset that wrongly removed high-id labels, and ensure clear_border raises ValueError on mask shape mismatch.
- Align MS-SSIM Gaussian kernel width with kernel_size and compute per-image reductions for batched inputs, matching torchmetrics’ behavior for default and non-default kernels.
- Compute average precision false positives and false negatives from distinct label counts instead of max label id so non-sequential labels with supplied matches are handled correctly.
- Allow get_true_features to accept vector morphology properties by validating against base feature names instead of expanded column names.
- Honor use_max_nyquist for index-unit radial bin edges so anisotropic shapes extend to the maximum axis Nyquist frequency.
- Exclude background label 0 from mesh-based feature extraction to match voxel.regionprops semantics.
- Center show_image_error’s color limits symmetrically around zero based on bit depth.

Enhancements:
- Introduce a shared GPU array detector and update SciPy and scikit-image proxies to inspect all arguments, correctly routing GPU arrays passed under any parameter name and safely coercing stray GPU arrays on the CPU path.

Tests:
- Add regression tests for corner cropping and odd-difference padding round trips, deconvolution observers and 2D inputs, GPU dispatch in SciPy and skimage proxies, MS-SSIM batch behavior and kernel-size handling, non-sequential-label average precision, vector feature handling, radial edge Nyquist behavior, mesh feature background exclusion, clear_border error typing, CUDA GPU-array detection, and symmetric error-map color limits.

### CODE REVIEW

File: cubic/plot_utils.py [hunk 1]
```diff
@@ -37,6 +37,6 @@ def show_image_error(
 ) -> Figure:
     """Display error map between teo images."""
     fig = plt.figure(figsize=figsize)
-    plt.imshow(img, cmap=cmap, vmin=(-(2**bit_depth) - 1), vmax=(2**bit_depth - 1))
+    plt.imshow(img, cmap=cmap, vmin=-(2**bit_depth - 1), vmax=(2**bit_depth - 1))
```
[the remark this answers is not in the document]
Comment (alxndrkalinin) [reply cubic/plot_utils.py]: Fixed in c3cab3e — corrected "teo" to "two".

File: cubic/cuda.py [hunk 2]
```diff
@@ -84,6 +84,53 @@ def get_device(array: np.ndarray) -> str:
...
 
+def is_gpu_array(array: Any) -> bool:
+    """Return True if ``array`` is a GPU array (CuPy ndarray or CUDA torch tensor).
+
+    Used by the SciPy / scikit-image proxies to detect the target device across
+    *all* arguments rather than only the first positional one. Uses the same
+    strict ``device`` test as :func:`asnumpy` (a ``cupy.cuda.Device`` exposes
+    ``id``; NumPy 2.x arrays report ``device == "cpu"``), so a non-array object
+    that merely carries a ``device`` attribute is not misclassified.
+    """
```
[the remark this answers is not in the document]
Comment (alxndrkalinin) [reply cubic/cuda.py:97]: Fixed in c3cab3e — reworded the docstring to describe the actual device-based heuristic (CuPy `Device`/CUDA tensor → GPU; numpy `"cpu"` or no `device` → CPU; a non-`"cpu"` string `device` → GPU), making clear this is a heuristic rather than a type check.

File: cubic/image_utils.py [hunk 3]
```diff
@@ -222,19 +230,16 @@ def crop_corner(
...
-        if "r" in corner and axis == axes[-1]:
-            slices[axis] = slice(-end if end is not None else None, None)
-        elif "b" in corner and axis == axes[-2]:
-            slices[axis] = slice(-end if end is not None else None, None)
+        size = min(size, img.shape[axis])
+        # "b"/"r" crop from the far edge (last/second-to-last axis); everything
+        # else (incl. "t"/"l") crops from the near edge.
+        crop_from_end = (axis == axes[-1] and "r" in corner) or (
+            axis == axes[-2] and "b" in corner
+        )
```
[the remark this answers is not in the document]
Comment (alxndrkalinin) [reply cubic/image_utils.py:238]: Fixed in bbc717e. The `and` evaluated `axis == axes[-2]` unconditionally, so a single-axis call raised `IndexError` even for corners the old code handled ("tl"/"tr"/"br"). Now guarded with `len(axes) >= 2` before consulting `axes[-2]`, preserving the prior single-axis behaviour.

File: tests/test_skimage_proxy.py [hunk 4]
```diff
@@ -4,7 +4,35 @@
...
+        gpu_res = mc_skimage.transform.rescale(
+            image=ascupy(img), scale=0.5, preserve_range=True
+        )
+        assert get_device(gpu_res) == "GPU"
+        assert np.allclose(asnumpy(gpu_res), cpu_res, atol=1e-5)
+    else:
+        assert np.allclose(
+            mc_skimage.transform.rescale(image=img, scale=0.5, preserve_range=True),
+            cpu_res,
+        )
```
[the remark this answers is not in the document]
Comment (alxndrkalinin) [reply tests/test_skimage_proxy.py]: Good catch — fixed in 31f7fb7. Switched to `measure.label(label_image=...)`, a keyword the old first-arg/`image`-only detection never inspected, so the test now genuinely fails on the pre-fix behaviour. Verified on the A40 (GPU branch runs, not skipped).

File: cubic/metrics/ms_ssim.py [hunk 5]
```diff
@@ -167,7 +182,6 @@ def ms_ssim(
     betas: tuple[float, ...] = DEFAULT_BETAS,
     kernel_size: int = 11,
     sigma: float = 1.5,
-    truncate: float = 3.5,
     K1: float = 0.01,
```
[the remark this answers is not in the document]
Comment (alxndrkalinin) [reply cubic/metrics/ms_ssim.py:185]: Intentional. `truncate` was redundant with `kernel_size` and was the direct cause of the kernel/pad desync this PR fixes — the Gaussian radius is now derived from `kernel_size` alone (`truncate = pad / sigma`). Re-accepting it as an ignored kwarg would silently discard caller intent, which the project avoids; a clear `TypeError` is preferable. As a pre-1.0 alpha the signature change is acceptable. Happy to add a deprecation shim if back-compat is desired.

File: cubic/metrics/ms_ssim.py [hunk 6]
```diff
@@ -96,23 +113,24 @@ def _torchmetrics_ssim_update(
...
+        spatial axes only — a scalar for ``(H, W)`` input or shape ``(N,)``
+        for ``(N, H, W)`` input, so callers can reduce per image rather
+        than pooling the whole batch. The CS map is averaged over an
+        additional ``[pad:-pad, pad:-pad]`` crop (``torchmetrics ssim.py:177``).
     """
     pad = (kernel_size - 1) // 2
-    # Pad only the spatial axes (last two); leave the batch axis alone.
+    # Tie skimage's Gaussian radius to ``pad`` so the kernel is exactly
+    # ``kernel_size`` wide regardless of ``kernel_size`` (radius == pad).
+    truncate = pad / sigma if sigma > 0 else 0.0
```
[the remark this answers is not in the document]
Comment (alxndrkalinin) [reply cubic/metrics/ms_ssim.py:132]: Fixed in 51dd5c9 — `ms_ssim` now rejects non-finite or non-positive `sigma` with a `ValueError`, matching the existing kernel_size/data_range validation.

File: cubic/segmentation/segment_utils.py [hunk 7]
```diff
@@ -283,24 +283,28 @@ def clear_xy_borders(label_image: np.ndarray, buffer_size: int = 0) -> np.ndarra
...
-                )
-                exclude_masks += [mask_idx] + list(overlap_masks)
+        if mask_idx in exclude_masks:
+            continue
+        binary_mask = label_image == mask_idx
+        dilated_mask = morphology.binary_dilation(binary_mask, morphology.cube(3))
+        mask_outline = dilated_mask & ~binary_mask
+
+        neighbor_ids = np.unique(label_image[mask_outline])
+        neighbor_ids = neighbor_ids[neighbor_ids != 0]
```
[the remark this answers is not in the document]
Comment (alxndrkalinin) [reply cubic/segmentation/segment_utils.py]: Fixed in 1818e95 — `exclude_masks` is now a `set[int]` (add/update), giving O(1) membership and removing duplicate work in the zeroing pass. Behaviour unchanged.

File: tests/test_scipy_proxy.py [hunk 8]
```diff
@@ -4,7 +4,31 @@
...
+@pytest.mark.parametrize("use_gpu", [False, True])
+def test_dispatch_on_array_passed_by_keyword(
+    use_gpu: bool, gpu_available: bool
+) -> None:
+    """A GPU array under a non-``input`` keyword routes to the GPU backend.
+
+    ``fftconvolve(in1=..., in2=...)`` has no ``input`` argument, so the old
+    first-arg-only detection saw no device, routed to CPU SciPy, and crashed
+    on the raw CuPy arrays. Detection now scans every argument.
+    """
```
[the remark this answers is not in the document]
Comment (alxndrkalinin) [reply tests/test_scipy_proxy.py:22]: Added in 31f7fb7 — `test_gpu_fallback_to_cpu_when_cupyx_unavailable` monkeypatches the cupyx import to fail for a GPU input and asserts the proxy warns, computes on CPU, and moves the result back to the GPU. Verified on the A40.
